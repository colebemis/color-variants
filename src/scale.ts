import * as Color from 'color'

interface Options {
  base: string
  steps: number
  lighten?: number
  hueShift?: number
  saturate?: number
}

export function scale({ base, steps, lighten = 1, hueShift = 0, saturate = 0 }: Options) {
  if (steps !== parseInt(steps.toString(), 10) || steps < 0) {
    throw new Error('The "steps" argmuent must be an positive integer.')
  }

  if (typeof lighten !== 'number' || lighten > 1 || lighten < -1) {
    throw new Error('The "lighten" argument must be a number between 1 and -1.')
  }

  if (typeof hueShift !== 'number') {
    throw new Error('The "hueShift" argument must be a number.')
  }

  if (typeof saturate !== 'number' || saturate > 1 || saturate < -1) {
    throw new Error('The "saturate" argument must be a number between 1 and -1.')
  }

  const color = Color(base)
  const colors = []

  for (let step = 1; step <= steps; step++) {
    const mixinColor = Color(lighten > 0 ? 'white' : 'black')
    colors.push(
      color
        .rotate((hueShift * step) / steps)
        .saturate((saturate * step) / steps)
        .mix(mixinColor, (Math.abs(lighten) * step) / steps)
        .hex()
    )
  }

  return colors
}

export default scale
