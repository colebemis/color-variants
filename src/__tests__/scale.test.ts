import scale from '../scale'

it('handles hex color values', () => {
  expect(scale({ base: '#ffaaaa', steps: 4 })).toMatchSnapshot()
})

it('handles rgb color values', () => {
  expect(scale({ base: 'rgb(0, 0, 255)', steps: 4 })).toMatchSnapshot()
})

it('handles hsl color values', () => {
  expect(scale({ base: 'hsl(0, 100%, 50%)', steps: 4 })).toMatchSnapshot()
})

it('handles color names', () => {
  expect(scale({ base: 'red', steps: 4 })).toMatchSnapshot()
})

it('handles positive "lighten" values', () => {
  expect(
    scale({ base: '#f00', steps: 4, lighten: 0.5, hueShift: 20, saturate: -0.5 })
  ).toMatchSnapshot()
})

it('handles negative "lighten" values', () => {
  expect(
    scale({ base: '#f00', steps: 4, lighten: -0.5, hueShift: 20, saturate: 0.5 })
  ).toMatchSnapshot()
})

it('throws if "base" is an invalid color value', () => {
  expect(() => scale({ base: 'foo', steps: 4 })).toThrowErrorMatchingSnapshot()
})

it('throws if "steps" is not an integer', () => {
  expect(() => scale({ base: '#f00', steps: 0.5 })).toThrowErrorMatchingSnapshot()
})

it('throws if "steps" is less than 0', () => {
  expect(() => scale({ base: '#f00', steps: -1 })).toThrowErrorMatchingSnapshot()
})

it('throws if "lighten" is greater than 1', () => {
  expect(() => scale({ base: '#f00', steps: 4, lighten: 2 })).toThrowErrorMatchingSnapshot()
})

it('throws if "lighten" is less than -1', () => {
  expect(() => scale({ base: '#f00', steps: 4, lighten: -2 })).toThrowErrorMatchingSnapshot()
})

it('throws if "saturate" is greater than 1', () => {
  expect(() => scale({ base: '#f00', steps: 4, saturate: 2 })).toThrowErrorMatchingSnapshot()
})

it('throws if "saturate" is less than -1', () => {
  expect(() => scale({ base: '#f00', steps: 4, saturate: -2 })).toThrowErrorMatchingSnapshot()
})
