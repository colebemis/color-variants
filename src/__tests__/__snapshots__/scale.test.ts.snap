// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles color names 1`] = `
Array [
  "#FF4040",
  "#FF8080",
  "#FFBFBF",
  "#FFFFFF",
]
`;

exports[`handles hex color values 1`] = `
Array [
  "#FFBFBF",
  "#FFD5D5",
  "#FFEAEA",
  "#FFFFFF",
]
`;

exports[`handles hsl color values 1`] = `
Array [
  "#FF4040",
  "#FF8080",
  "#FFBFBF",
  "#FFFFFF",
]
`;

exports[`handles negative "lighten" values 1`] = `
Array [
  "#DF0700",
  "#BF1000",
  "#9F1900",
  "#802000",
]
`;

exports[`handles positive "lighten" values 1`] = `
Array [
  "#F13E2E",
  "#E77058",
  "#E1967E",
  "#DFB59F",
]
`;

exports[`handles rgb color values 1`] = `
Array [
  "#4040FF",
  "#8080FF",
  "#BFBFFF",
  "#FFFFFF",
]
`;

exports[`throws if "base" is an invalid color value 1`] = `"Unable to parse color from string: foo"`;

exports[`throws if "lighten" is greater than 1 1`] = `"The \\"lighten\\" argument must be a number between 1 and -1."`;

exports[`throws if "lighten" is less than -1 1`] = `"The \\"lighten\\" argument must be a number between 1 and -1."`;

exports[`throws if "saturate" is greater than 1 1`] = `"The \\"saturate\\" argument must be a number between 1 and -1."`;

exports[`throws if "saturate" is less than -1 1`] = `"The \\"saturate\\" argument must be a number between 1 and -1."`;

exports[`throws if "steps" is less than 0 1`] = `"The \\"steps\\" argmuent must be an positive integer."`;

exports[`throws if "steps" is not an integer 1`] = `"The \\"steps\\" argmuent must be an positive integer."`;
